File: server/src/domain/pipeline.rs
//! Wire types for the per-element processing pipeline admin API
//! (`/api/admin/pipeline`) and the `pipeline.stats` live event. Pure data (serde
//! + ts-rs); the engine that produces them lives in
//! [`crate::services::pipeline`], persistence in [`crate::db::pipeline`].
//!
//! A *stage* (probe, metadata, storyboard, markers, embed) processes one subject
//! (a file / item / show / season) at a time and records the outcome in the
//! `pipeline_tasks` ledger, so a re-run only does the non-done work and failures
//! are individually visible + retriable.

use serde::Serialize;
use ts_rs::TS;

/// Health counters for one pipeline stage, aggregated from the ledger. Carried by
/// both the REST view and the throttled `pipeline.stats` WS event, so the
/// dashboard updates live without polling per-task rows.
#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct StageStat {
    /// Short stage key (`"probe"`, `"markers"`, …); i18n base `pipeline.stage.{stage}`.
    pub stage: String,
    /// Full job key of the stage's drain job (`"pipeline.probe"`), to correlate
    /// with the existing `/api/admin/jobs` run/schedule/log surface.
    pub key: String,
    /// What one task operates on: `"file" | "item" | "show" | "season"`.
    pub subject_kind: String,
    pub pending: i64,
    pub running: i64,
    pub done: i64,
    pub failed: i64,
    pub blocked: i64,
}

/// `GET /api/admin/pipeline`: every stage's health, in DAG order, plus whether the
/// whole pipeline is currently held by the global admin pause.
#[derive(Debug, Clone, Serialize, TS)]
#[ts(export)]
pub struct PipelineView {
    pub stages: Vec<StageStat>,
    pub paused: bool,
}

/// The status of one treatment (stage) as applied to a single catalog element,
/// for the per-element "Traitements" panel on a film/episode/show page.
#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct Treatment {
    /// Short stage key (`"probe"`, `"metadata"`, `"storyboard"`, `"markers"`, `"embed"`).
    pub key: String,
    /// `"done" | "missing" | "pending" | "running" | "failed"`.
    pub status: String,
    /// Failure message when `status == "failed"` (only populated in the elements
    /// list, for the detail drawer). `None` elsewhere.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Per-series aggregate over its episodes, for the elements list (the client
/// formats the localized "28 episodes · probed 28/28 · …" line).
#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct EpStats {
    pub episodes: i64,
    pub probed: i64,
    pub storyboarded: i64,
    pub seasons: i64,
    pub marker_seasons: i64,
}

/// One catalog element (film / series / episode) with the status of each
/// treatment applied to it and an overall roll-up, for the pipeline elements list.
#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ElementRow {
    pub id: String,
    /// `"film" | "series" | "episode"`.
    pub kind: String,
    pub title: String,
    /// Cached poster URL (`/api/images/…`) resolved from TMDB metadata; `None`
    /// falls back to a placeholder client-side. Episodes borrow their show's.
    pub poster: Option<String>,
    /// Structured hints so the client builds a localized subtitle.
    pub year: Option<u32>,
    pub genre: Option<String>,
    pub duration_ms: Option<u64>,
    pub season_count: Option<u32>,
    pub treatments: Vec<Treatment>,
    /// `"ok" | "pending" | "running" | "failed"`.
    pub overall: String,
    /// Series only: per-episode aggregate for the detail drawer.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ep_stats: Option<EpStats>,
}

/// Status tally over ALL elements (unfiltered), for the filter chips + header.
#[derive(Debug, Clone, Default, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ElementCounts {
    pub total: i64,
    pub ok: i64,
    pub pending: i64,
    pub running: i64,
    pub failed: i64,
    pub film: i64,
    pub series: i64,
    pub episode: i64,
}

/// `GET /api/admin/pipeline/elements`: a filtered, paginated page of the catalog
/// with per-element treatment state, plus the full-catalog counts for the chips.
#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct PipelineElements {
    pub total: i64,
    pub page: i64,
    pub pages: i64,
    pub counts: ElementCounts,
    pub elements: Vec<ElementRow>,
}

/// `GET /api/admin/pipeline/item|show/:id`: every treatment that applies to this
/// element + whether it has been done, derived from the real artifacts (probed
/// flag, metadata, cached sheet, markers, vector) with the ledger overlaid for
/// in-progress / failed states.
#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct ElementProcessing {
    pub treatments: Vec<Treatment>,
}

/// One failed (or otherwise notable) ledger row, for the stage drill-down. The
/// human-readable `title` is resolved against the catalog by the API layer.
#[derive(Debug, Clone, Serialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct PipelineTaskView {
    pub stage: String,
    pub subject_kind: String,
    pub subject_id: String,
    /// Best-effort catalog title for the subject (falls back to the id).
    pub title: String,
    pub status: String,
    pub attempts: i64,
    pub error: Option<String>,
    pub finished_at: Option<i64>,
}

File: packages/client/src/schemas/pipeline.ts
// Runtime schemas for the pipeline / jobs / naming / VPN / LLM domain.
//
// Mirrors the ts-rs generated wire types (the Rust structs are the single source
// of truth) with runtime validation and branded ids. The `true satisfies
// SameKeys<…>` lines are compile-time drift guards: change a Rust struct + rerun
// `gen:types` and the build breaks here until the schema is updated. See
// `./accounts` for the template and `./drift` for how the guards work.

import { z } from 'zod';
import { EpStats } from './admin';
import { JobRunId } from './ids';
import { Category, Treatment } from './media';

// ── Jobs ────────────────────────────────────────────────────────────────────

/** One recorded execution of a job. */
export const JobRun = z.object({
  id: JobRunId,
  jobKey: z.string(),
  trigger: z.string(),
  status: z.string(),
  startedAt: z.number(),
  finishedAt: z.number().nullable(),
  durationMs: z.number().nullable(),
  progressDone: z.number().nullable(),
  progressTotal: z.number().nullable(),
  error: z.string().nullable(),
});
export type JobRun = z.infer<typeof JobRun>;

/** A job's definition + current state, as listed in the admin console. */
export const JobInfo = z.object({
  key: z.string(),
  name: z.string(),
  description: z.string(),
  category: Category,
  schedule: z.string().nullable(),
  defaultSchedule: z.string().nullable(),
  customized: z.boolean(),
  enabled: z.boolean(),
  running: z.boolean(),
  runId: JobRunId.nullable(),
  progressDone: z.number().nullable(),
  progressTotal: z.number().nullable(),
  nextRunAt: z.number().nullable(),
  lastRun: JobRun.nullable(),
});
export type JobInfo = z.infer<typeof JobInfo>;

/** One persisted log line of a run. */
export const JobLog = z.object({
  ts: z.number(),
  level: z.string(),
  message: z.string(),
});
export type JobLog = z.infer<typeof JobLog>;

/** `GET /api/admin/jobs/:key` a job plus its recent run history. */
export const JobDetail = z.object({
  info: JobInfo,
  runs: z.array(JobRun),
});
export type JobDetail = z.infer<typeof JobDetail>;

/** `GET /api/admin/jobs`. */
export const JobsView = z.object({
  jobs: z.array(JobInfo),
});
export type JobsView = z.infer<typeof JobsView>;

// ── Pipeline ──────────────────────────────────────────────────────────────

/** Health counters for one pipeline stage, aggregated from the ledger. */
export const StageStat = z.object({
  stage: z.string(),
  key: z.string(),
  subjectKind: z.string(),
  pending: z.number(),
  running: z.number(),
  done: z.number(),
  failed: z.number(),
  blocked: z.number(),
});
export type StageStat = z.infer<typeof StageStat>;

/** Status tally over ALL elements (unfiltered), for the filter chips + header. */
export const ElementCounts = z.object({
  total: z.number(),
  ok: z.number(),
  pending: z.number(),
  running: z.number(),
  failed: z.number(),
  film: z.number(),
  series: z.number(),
  episode: z.number(),
});
export type ElementCounts = z.infer<typeof ElementCounts>;

/** Every treatment that applies to an element + whether it has been done. */
export const ElementProcessing = z.object({
  treatments: z.array(Treatment),
});
export type ElementProcessing = z.infer<typeof ElementProcessing>;

/** One catalog element (film / series / episode) with per-treatment status. */
export const ElementRow = z.object({
  id: z.string(),
  kind: z.string(),
  title: z.string(),
  poster: z.string().nullable(),
  year: z.number().nullable(),
  genre: z.string().nullable(),
  durationMs: z.number().nullable(),
  seasonCount: z.number().nullable(),
  treatments: z.array(Treatment),
  overall: z.string(),
  epStats: EpStats.nullish(),
});
export type ElementRow = z.infer<typeof ElementRow>;

/** `GET /api/admin/pipeline/elements`: a filtered, paginated page of the catalog. */
export const PipelineElements = z.object({
  total: z.number(),
  page: z.number(),
  pages: z.number(),
  counts: ElementCounts,
  elements: z.array(ElementRow),
});
export type PipelineElements = z.infer<typeof PipelineElements>;

/** One failed (or otherwise notable) ledger row, for the stage drill-down. */
export const PipelineTaskView = z.object({
  stage: z.string(),
  subjectKind: z.string(),
  subjectId: z.string(),
  title: z.string(),
  status: z.string(),
  attempts: z.number(),
  error: z.string().nullable(),
  finishedAt: z.number().nullable(),
});
export type PipelineTaskView = z.infer<typeof PipelineTaskView>;

/** `GET /api/admin/pipeline`: every stage's health, in DAG order. */
export const PipelineView = z.object({
  stages: z.array(StageStat),
  paused: z.boolean(),
});
export type PipelineView = z.infer<typeof PipelineView>;

// ── Naming / organize ───────────────────────────────────────────────────────

/** Example rendered names for the live preview. */
export const SampleNames = z.object({
  movie: z.string(),
  episode: z.string(),
});
export type SampleNames = z.infer<typeof SampleNames>;

/** The five naming templates plus the global case transform. */
export const NamingTemplatesView = z.object({
  movieFolder: z.string(),
  movieFile: z.string(),
  seriesFolder: z.string(),
  seasonFolder: z.string(),
  episodeFile: z.string(),
  case: z.string(),
});
export type NamingTemplatesView = z.infer<typeof NamingTemplatesView>;

/** `GET /api/admin/organize/naming` current templates + a rendered sample. */
export const NamingView = z.object({
  templates: NamingTemplatesView,
  sample: SampleNames,
});
export type NamingView = z.infer<typeof NamingView>;

/** One file the rename tool would move. */
export const OrganizeMove = z.object({
  title: z.string(),
  kind: z.string(),
  from: z.string(),
  to: z.string(),
});
export type OrganizeMove = z.infer<typeof OrganizeMove>;

/** `GET /api/admin/organize/preview`. */
export const OrganizePlan = z.object({
  moves: z.array(OrganizeMove),
  totalFiles: z.number(),
  matching: z.number(),
});
export type OrganizePlan = z.infer<typeof OrganizePlan>;

/** `POST /api/admin/organize/apply` result. */
export const OrganizeResult = z.object({
  moved: z.number(),
  failed: z.number(),
  errors: z.array(z.string()),
});
export type OrganizeResult = z.infer<typeof OrganizeResult>;

/** `POST /api/admin/acquisition/analyze` body. */
export const AnalyzeBody = z.object({
  magnetOrUrl: z.string(),
});
export type AnalyzeBody = z.infer<typeof AnalyzeBody>;

// ── VPN ───────────────────────────────────────────────────────────────────

/** The kill switch's view of the tunnel. */
export const VpnStatusView = z.object({
  connected: z.boolean(),
  exitIp: z.string().nullable(),
  paused: z.boolean(),
});
export type VpnStatusView = z.infer<typeof VpnStatusView>;

/** `POST /api/admin/vpn/test` a live probe through (and around) the proxy. */
export const VpnTestResult = z.object({
  sealed: z.boolean(),
  proxiedIp: z.string().nullable(),
  directIp: z.string().nullable(),
  error: z.string().nullable(),
});
export type VpnTestResult = z.infer<typeof VpnTestResult>;

/** `GET /api/admin/vpn` the VPN configuration card's state. */
export const VpnAdminView = z.object({
  wgConfigured: z.boolean(),
  bridgeRunning: z.boolean(),
  localPort: z.number(),
  status: VpnStatusView.nullable(),
});
export type VpnAdminView = z.infer<typeof VpnAdminView>;

/** `PUT /api/admin/vpn` body. `wgConfig` is write-only. */
export const SaveVpnBody = z.object({
  wgConfig: z.string().nullable(),
  localPort: z.number().nullable(),
});
export type SaveVpnBody = z.infer<typeof SaveVpnBody>;

// ── LLM ───────────────────────────────────────────────────────────────────

/** One configured provider as shown to the admin (API key never returned). */
export const LlmProviderView = z.object({
  id: z.string(),
  name: z.string(),
  provider: z.string(),
  baseUrl: z.string(),
  model: z.string(),
  hasApiKey: z.boolean(),
  temperature: z.number(),
  maxTokens: z.number(),
  reasoning: z.boolean(),
});
export type LlmProviderView = z.infer<typeof LlmProviderView>;

/** `GET /api/admin/llm` the multi-provider LLM configuration. */
export const LlmAdminConfig = z.object({
  enabled: z.boolean(),
  defaultId: z.string(),
  providers: z.array(LlmProviderView),
});
export type LlmAdminConfig = z.infer<typeof LlmAdminConfig>;
